railties/lib/rails/command.rb

```diff
@@ -49,7 +49,12 @@ def invoke(full_namespace, args = [], **config)
         if command && command.all_commands[command_name]
           command.perform(command_name, args, config)
         else
-          find_by_namespace("rake").perform(full_namespace, args, config)
+          rake_namespace = find_by_namespace("rake")
+          if rake_namespace.performable_commands_and_options.include?(full_namespace)
+            rake_namespace.perform(full_namespace, args, config)
+          else
+            suggest_commands(full_namespace)
+          end
```

Comment:
I don't think preventing the exception is the right approach because it changes the contract of `Command.invoke`.  I tried an alternative approach in #39541.  I added you as a co-author to share credit for exploring this.

Reply:
Thanks for exploring this! I'm not a fan of of the `#performable_commands_and_options` approach either (too many corner cases I'm probably missing).

I've also submitted a PR to `Rake` to raise a new error if a task isn't found: https://github.com/ruby/rake/pull/354 This would allow us to make it work more like `Thor::Error` which also supresses backtraces for users. https://github.com/erikhuda/thor/blob/master/lib/thor/error.rb